### PR TITLE
Change Submaps to only track the two active Submaps.

### PR DESCRIPTION
**Pull request by SirVer**
Related to #283.

### CODE REVIEW

File: cartographer/mapping_3d/submaps.cc [hunk 1]
```diff
@@ -413,15 +398,17 @@ void Submaps::InsertRangeData(const sensor::RangeData& range_data,
...
+    Submap* submap = submaps_.front().get();
     CHECK(!submap->finished_);
     submap->finished_ = true;
+    ++matching_submap_index_;
+    submaps_.erase(submaps_.begin());
   }
   submaps_.emplace_back(new Submap(options_.high_resolution(),
                                    options_.low_resolution(), local_pose));
-  LOG(INFO) << "Added submap " << size();
+  LOG(INFO) << "Added submap " << matching_submap_index_ + 1;
```
Comment (wohe) [inline cartographer/mapping_3d/submaps.cc]: `matching_submap_index_ + submaps_.size()`?

File: cartographer/mapping_3d/sparse_pose_graph.cc [hunk 2]
```diff
@@ -424,6 +424,16 @@ SparsePoseGraph::GetTrajectoryNodes() {
   return trajectory_nodes_.data();
 }
 
+std::shared_ptr<const Submap> SparsePoseGraph::GetSubmap(
+    const mapping::SubmapId& submap_id) {
+  common::MutexLocker locker(&mutex_);
+  // Conceptually, this method should only be used to ask for finished submaps.
+  // But since we cannot guarantee that the optimization problem already ran,
+  // submap_data.state can be != kFinished.
+  const auto& submap_data = submap_data_.at(submap_id);
```
Comment (wohe) [inline cartographer/mapping_3d/sparse_pose_graph.cc]: Combine into one line?

File: cartographer/mapping_3d/sparse_pose_graph.cc [hunk 3]
```diff
@@ -424,6 +424,16 @@ SparsePoseGraph::GetTrajectoryNodes() {
   return trajectory_nodes_.data();
 }
 
+std::shared_ptr<const Submap> SparsePoseGraph::GetSubmap(
+    const mapping::SubmapId& submap_id) {
+  common::MutexLocker locker(&mutex_);
+  // Conceptually, this method should only be used to ask for finished submaps.
```
Comment (wohe) [inline cartographer/mapping_3d/sparse_pose_graph.cc]: Add a comment to the struct definition that `state` is about the state of the submap in the background processing?

File: cartographer/mapping_3d/optimizing_local_trajectory_builder.cc [hunk 4]
```diff
@@ -410,13 +410,13 @@ OptimizingLocalTrajectoryBuilder::InsertIntoSubmap(
     return nullptr;
   }
   std::vector<std::shared_ptr<const Submap>> insertion_submaps;
-  for (int insertion_index : submaps_->insertion_indices()) {
-    insertion_submaps.push_back(submaps_->Get(insertion_index));
+  for (std::shared_ptr<Submap> ptr : active_submaps_.submaps()) {
```
Comment (wohe) [inline cartographer/mapping_3d/optimizing_local_trajectory_builder.cc]: `ptr` -> `submap`?

File: cartographer/mapping_3d/kalman_local_trajectory_builder.cc [hunk 5]
```diff
@@ -220,10 +220,10 @@ KalmanLocalTrajectoryBuilder::InsertIntoSubmap(
     return nullptr;
   }
   std::vector<std::shared_ptr<const Submap>> insertion_submaps;
-  for (int insertion_index : submaps_->insertion_indices()) {
-    insertion_submaps.push_back(submaps_->Get(insertion_index));
+  for (std::shared_ptr<Submap> ptr : active_submaps_.submaps()) {
```
Comment (wohe) [inline cartographer/mapping_3d/kalman_local_trajectory_builder.cc]: Same here: `submap`?

File: cartographer/mapping_3d/global_trajectory_builder.cc [hunk 6]
```diff
@@ -36,9 +36,23 @@ int GlobalTrajectoryBuilder::num_submaps() {
 
 GlobalTrajectoryBuilder::SubmapData GlobalTrajectoryBuilder::GetSubmapData(
     const int submap_index) {
-  return {local_trajectory_builder_->submaps()->Get(submap_index),
-          sparse_pose_graph_->GetSubmapTransform(
-              mapping::SubmapId{trajectory_id_, submap_index})};
+  auto* active_submaps = local_trajectory_builder_->active_submaps();
```
Comment (wohe) [inline cartographer/mapping_3d/global_trajectory_builder.cc]: For now directly ask the `sparse_pose_graph_` which is guaranteed to know about the submap?

File: cartographer/mapping_3d/sparse_pose_graph.cc (hunk 3 above)
Comment (SirVer) [inline cartographer/mapping_3d/sparse_pose_graph.cc]: is already there.

File: cartographer/mapping_2d/submaps.h [hunk 7]
```diff
@@ -90,7 +90,7 @@ class ActiveSubmaps {
   void AddSubmap(const Eigen::Vector2f& origin);
 
   const proto::SubmapsOptions options_;
-  int matching_submap_index_;
+  int matching_submap_index_ = 0;
```
Comment (wohe) [inline cartographer/mapping_2d/submaps.h]: Drop from the initializer list?

**Pull request merged**